kimchi/src/tests/chacha.rs

```diff
@@ -98,3 +99,159 @@ fn chacha_prover() {
+            typ,
+            coeffs: vec![],
+            wires: Wire::new(i),
+        })
+        // Pad with generic gates to get a sufficiently-large domain.
+        .chain((4..513).map(|i| CircuitGate {
+            typ: GateType::Generic,
+            coeffs: vec![Fp::zero(); 10],
+            wires: Wire::new(i),
+        }))
+        .collect();
+
+    let mut rows = vec![];
+
+    let s: Vec<u32> = vec![
+        0x61707865, 0x3320646e, 0x79622d32, 0x6b206574, 0x03020100, 0x07060504, 0x0b0a0908,
+        0x0f0e0d0c, 0x13121110, 0x17161514, 0x1b1a1918, 0x1f1e1d1c, 0x00000001, 0x09000000,
+        0x4a000000, 0x00000000,
+    ];
+
+    let mut fakes = vec![vec![]; 3];
+
+    // Push the witnesses for a gate. When `fake` is true, the XOR operation is faked to return 0.
+    // This has been shamelessly copied from the chacha module and vandalised.
+    let mut push_rows = |fake: bool| {
+        let k = 16;
+        let x = 0;
+        let y = 1;
+        let z = 2;
+        let f = |t: u32| Fp::from(t);
+        let nyb = |t: u32, i: usize| f((t >> (4 * i)) & 0b1111);
+
+        let top_bit = (((s[x] as u64) + (s[z] as u64)) >> 32) as u32;
+        let xprime = u32::wrapping_add(s[x], s[z]);
+        let y_xor_xprime = if fake {
+            for nybble in 0..8 {
+                fakes[0].push(nyb(0, nybble));
+                fakes[1].push(nyb(xprime, nybble));
+                fakes[2].push(nyb(s[y], nybble));
+            }
+            0
+        } else {
+            s[y] ^ xprime
+        };
+        let yprime = y_xor_xprime.rotate_left(k);
+
+        rows.push(vec![
+            f(s[x]),
+            f(s[y]),
+            f(s[z]),
+            nyb(y_xor_xprime, 0),
+            nyb(y_xor_xprime, 1),
+            nyb(y_xor_xprime, 2),
+            nyb(y_xor_xprime, 3),
+            nyb(xprime, 0),
+            nyb(xprime, 1),
+            nyb(xprime, 2),
+            nyb(xprime, 3),
+            nyb(s[y], 0),
+            nyb(s[y], 1),
+            nyb(s[y], 2),
+            nyb(s[y], 3),
+        ]);
+        rows.push(vec![
+            f(xprime),
+            f(yprime),
+            f(top_bit),
+            nyb(y_xor_xprime, 4),
+            nyb(y_xor_xprime, 5),
+            nyb(y_xor_xprime, 6),
+            nyb(y_xor_xprime, 7),
+            nyb(xprime, 4),
+            nyb(xprime, 5),
+            nyb(xprime, 6),
+            nyb(xprime, 7),
+            nyb(s[y], 4),
+            nyb(s[y], 5),
+            nyb(s[y], 6),
+            nyb(s[y], 7),
+        ]);
+    };
+
+    // One real witness..
+    push_rows(false);
+    // .. and one fake witness.
+    push_rows(true);
+
+    let lookup_tables = vec![
+        LookupTable {
+            id: 0,
+            data: vec![vec![Fp::from(0); 10]],
+        },
+        LookupTable {
+            id: table_id,
+            data: fakes,
+        },
+    ];
+
+    // create the index
+    let index = new_index_for_test_with_lookups(gates, PUBLIC, lookup_tables);
```

Comment:
I'm not sure I understand how this could even work, since the verifier will not use these tables when creating the linearized commitment of the lookup grand product argument

Reply:
ok actually nevermind, the verifier index will copy the big lookup table from the prover index